cluster/v1alpha1/helpers.go

```diff
@@ -0,0 +1,311 @@
+	if err != nil {
+		return &strategy, result, err
+	}
+	restclusters := r.pdTracker.ExistingBesides(clusterGroupKeys).UnsortedList()
+
+	result = progressivePerCluster(restclusters, length, failureTimeout, statusFunc)
+	return &strategy, result, nil
+}
+
+func (r *RolloutHandler) getProgressivePerGroupClusters(rolloutStrategy RolloutStrategy, statusFunc ClusterRolloutStatusFunc) (*RolloutStrategy, map[string]ClusterRolloutStatus, error) {
+	strategy := RolloutStrategy{Type: ProgressivePerGroup}
+	strategy.ProgressivePerGroup = rolloutStrategy.ProgressivePerGroup.DeepCopy()
+	if strategy.ProgressivePerGroup == nil {
+		strategy.ProgressivePerGroup = &RolloutProgressivePerGroup{}
+	}
+
+	// upgrade madatory decision groups first
+	groupKeys := decisionGroupsToGroupKeys(strategy.ProgressivePerGroup.MandatoryDecisionGroups.MandatoryDecisionGroups)
+	clusterGroupKeys, clusterGroups := r.pdTracker.ExistingClusterGroups(groupKeys)
+
+	result := progressivePerGroup(clusterGroupKeys, clusterGroups, maxTimeDuration, statusFunc)
+	if len(result) > 0 {
+		return &strategy, result, nil
+	}
+
+	// parse timeout
+	failureTimeout, err := parseTimeout(strategy.ProgressivePerGroup.Timeout.Timeout)
+	if err != nil {
+		return &strategy, result, err
+	}
+
+	// upgrade the rest decision groups
+	restClusterGroupKeys, restClusterGroups := r.pdTracker.ExistingClusterGroupsBesides(clusterGroupKeys)
+
+	result = progressivePerGroup(restClusterGroupKeys, restClusterGroups, failureTimeout, statusFunc)
+	return &strategy, result, nil
+}
+
+func progressivePerCluster(clusters []string, length int, timeout time.Duration, statusFunc ClusterRolloutStatusFunc) map[string]ClusterRolloutStatus {
+	result := map[string]ClusterRolloutStatus{}
+	if length == 0 {
+		return result
+	}
+
+	// sort the clusters
+	sort.Strings(clusters)
+	for _, cluster := range clusters {
+		status := statusFunc(cluster)
+		if needToUpdate(status, timeout) {
+			result[cluster] = status
+		}
+
+		if (len(result)%length == 0) && len(result) > 0 {
+			return result
+		}
+	}
+
+	return result
+}
+
+func progressivePerGroup(clusterGroupKeys []clusterv1beta1.GroupKey, clusterGroups map[clusterv1beta1.GroupKey]sets.Set[string], timeout time.Duration, statusFunc ClusterRolloutStatusFunc) map[string]ClusterRolloutStatus {
+	result := map[string]ClusterRolloutStatus{}
+
+	for _, key := range clusterGroupKeys {
+		if subclusters, ok := clusterGroups[key]; ok {
+			for _, cluster := range subclusters.UnsortedList() {
+				status := statusFunc(cluster)
+				if needToUpdate(status, timeout) {
+					result[cluster] = status
+				}
+			}
+
+			if len(result) > 0 {
+				return result
+			}
+		}
+	}
+
+	return result
+}
+
+func needToUpdate(status ClusterRolloutStatus, timeout time.Duration) bool {
+	switch status.Status {
+	case ToApply:
+		return true
+	case Progressing:
+		return true
+	case Succeed:
+		return false
+	case Failed:
```

Comment:
I believe we need to distinguish between timeout and failed. It can be failed before reaching timeout AND as well it can be timed out but not failed  (as the status will continue to reconcile it will reach failed/success later).

Reply:
Sounds good. Failed or progressing is the status provided by consumer contoller's ClusterRolloutStatusFunc, and should return timeout in the result of ClusterRolloutStatus by the RolloutHandler.

Reply:
Have updated the func, when the cluster staus is progressing or failed and reach timeout, will return status TimeOut.